upgrade/v1beta1/upgrade.go

```diff
@@ -0,0 +1,210 @@
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+" failed")
+	}()
+
+	lhClientV1beta1, err := lhclientset.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "unable to get clientset for v1beta1")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhornV1beta1.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return errors.Wrap(err, "unable to create scheme for v1beta1")
+	}
+
+	if err := upgradeVolumes(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeEngineImages(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeBackupTargets(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeNodes(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+
+	logrus.Infof("%v: completed", upgradeLogPrefix)
+	return nil
+}
+
+func copyConditions(fromConditions map[string]longhornV1beta1.Condition) ([]longhorn.Condition, error) {
+	toConditions := []longhorn.Condition{}
+	for _, fromCon := range fromConditions {
+		toCon := longhorn.Condition{}
+		if err := copier.Copy(&toCon, &fromCon); err != nil {
+			return nil, err
+		}
+		toConditions = append(toConditions, toCon)
+	}
+	return toConditions, nil
+}
+
+func CanUpgrade(config *restclient.Config, namespace string) (bool, error) {
+	lhClient, err := lhclientset.NewForConfig(config)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to get clientset for v1beta1")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhornV1beta1.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return false, errors.Wrap(err, "unable to create scheme for v1beta1")
+	}
+
+	setting, err := lhClient.LonghornV1beta1().Settings(namespace).Get(context.TODO(), string(types.SettingNameDefaultEngineImage), metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("setting %v not found", string(types.SettingNameDefaultEngineImage))
+			return true, nil
+		}
+
+		return false, errors.Wrap(err, fmt.Sprintf("unable to get setting %v", string(types.SettingNameDefaultEngineImage)))
+	}
+
+	return false, errors.Wrapf(err, "unable to upgrade from %v directly", setting.Value)
+}
+
+func upgradeVolumes(namespace string, lhClientV1beta1 *lhclientset.Clientset, lhClient *lhclientset.Clientset) error {
+	volumes, err := lhClientV1beta1.LonghornV1beta1().Volumes(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "unable to list volumes")
+	}
+	for _, old := range volumes.Items {
+		new := &longhorn.Volume{}
+
+		if err := copier.Copy(&new, &old); err != nil {
+			return err
+		}
+
+		conditions, err := copyConditions(old.Status.Conditions)
+		if err != nil {
+			return errors.Wrap(err, "upgrade: copy volume conditions")
+		}
+		new.Status.Conditions = conditions
+
+		new, err = lhClient.LonghornV1beta2().Volumes(namespace).UpdateStatus(context.TODO(), new, metav1.UpdateOptions{})
+		if err != nil {
```

Comment:
Should this part check the conflicts as well?

Reply:
Always return error if there is any error and conflict and retry next time.